File: Advanced Methods of Programming/CSharp Toy Language Interpreter/ToyLanguageInterpreter/Src/View/TextMenu.cs
using System;
using System.Collections.Generic;
using ToyLanguageInterpreter.Src.Model.Commands;

namespace ToyLanguageInterpreter.Src.View
{
    class TextMenu
    {
        private Dictionary<string, Command> commands; 

        public TextMenu()
        {
            commands = new Dictionary<string, Command>();
        }

        public void AddCommand(Command c)
        {
            commands.Add(c.Key, c);
        }

        private void PrintMenu()
        {
            foreach (Command com in commands.Values)
            {
                string line = string.Format("{0} : {1}", com.Key, com.Description);
                Console.WriteLine(line);
            }
        }

        public void Show()
        {
            while (true)
            {
                PrintMenu();
                Console.Write("Input the option: ");

                string key = Console.ReadLine();

                Command com;

                if (commands.TryGetValue(key, out com))
                {
                    com.Execute();
                }
                else
                {
                    Console.WriteLine("Invalid Option");
                }
            }
        }
    }
}

File: Advanced Methods of Programming/Toy Language Interpreter/src/View/CLI/TextMenu.java
package View.CLI;

import Model.Commands.Command;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class TextMenu {

    private Map<String, Command> commands;
    public TextMenu(){
        commands=new HashMap<>();
    }

    public void addCommand(Command c){
        commands.put(c.getKey(),c);
    }

    private void printMenu(){
        for(Command com : commands.values()){
            String line=String.format("%4s : %s", com.getKey(), com.getDescription());
            System.out.println(line);
        }
    }
    public void show(){
        Scanner scanner=new Scanner(System.in);

        while(true){
            //prompt user with possibilities
            printMenu();
            System.out.printf("Input the option: ");

            //wait user input
            String key=scanner.nextLine();

            //execute command given
            Command com = commands.get(key);

            if (com==null){
                System.out.println("Invalid Option");
                continue;
            }

            com.execute();
        }
    }
}